### PR TITLE
MSCCL: add support for out-of-place all reduce

### PR DESCRIPTION
**Pull request by wenkaidu**
## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** _"Internal", or link to GitHub issue (if applicable)._

**What were the changes?**  
add support for out-of-place all reduce to MSCCL

**Why were the changes made?**  
Improve out-of-place all reduce performance

**How was the outcome achieved?**  
add support for out-of-place all reduce to MSCCL

**Additional Documentation:**  

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Review by PedramAlizadeh — APPROVED** [review]

File: src/device/msccl_kernel_impl.h [hunk 1]
```diff
@@ -372,7 +372,7 @@ __device__ __forceinline__ void mscclRunInterpreter(
           return;
       }
       if (t->hasDependence && tid == nthreads-1)
-        __atomic_store_n(&mscclFlags[bid].flag, (uint64_t) COMPUTE_FLAG(workIndex, iter, step), ((t->type == MSCCL_REDUCE || t->type == MSCCL_RECV) && (t->dstBuffer != MSCCL_SCRATCH_BUFFER)) ? __ATOMIC_RELEASE : __ATOMIC_RELAXED);
```
Comment (nusislam) [inline src/device/msccl_kernel_impl.h:375]: Should a comment be added here explaining this change?

Comment (wenkaidu) [reply src/device/msccl_kernel_impl.h:375]: Sure. Out of place all reduce adds sending from output buffer, thus need cache flush/invalidate when this happens

**Review by nusislam — APPROVED** [review]

File: src/device/msccl_kernel_impl.h (hunk 1 above)
Comment (nusislam) [reply src/device/msccl_kernel_impl.h:375]: The PR looks good. There are some test failures that need to be checked before merging.

Comment (wenkaidu) [reply src/device/msccl_kernel_impl.h:375]: I guess you mean gfx906 failures? These have been there for months. I think re-run will pass

**Pull request merged**